luatest/cluster.lua

```diff
@@ -0,0 +1,349 @@
+end
+
+--- Stop the whole cluster.
+--
+-- @tab self Cluster itself.
+local function cluster_stop(self)
+    for _, iserver in ipairs(self._servers or {}) do
+        iserver:stop()
+    end
+end
+
+--- Drop the cluster's servers.
+--
+-- @tab self Cluster itself.
+local function cluster_drop(self)
+    for _, iserver in ipairs(self._servers or {}) do
+        iserver:drop()
+    end
+    self._servers = nil
+    self._server_map = nil
+end
+
+--- Sync the cluster object with the new config.
+--
+-- It performs the following actions.
+--
+-- * Write the new config into the config file.
+-- * Update the internal list of instances.
+--
+-- @tab self Cluster itself.
+-- @tab config New config.
+local function cluster_sync(self, config)
+    assert(type(config) == 'table')
+
+    local instance_names = instance_names_from_config(config)
+
+    treegen.write_file(self._dir, self._config_file_rel, yaml.encode(config))
+
+    for i, name in ipairs(instance_names) do
+        if self._server_map[name] == nil then
+            local iserver = server:new(fun.chain(self._server_opts, {
+                alias = name,
+            }):tomap())
+            table.insert(self._servers, i, iserver)
+            self._server_map[name] = iserver
+        end
+    end
+
+end
+
+--- Reload configuration on all the instances.
+--
+-- @tab self Cluster itself.
+-- @tab[opt] config New config.
+local function cluster_reload(self, config)
+    assert(config == nil or type(config) == 'table')
+
+    -- Rewrite the configuration file if a new config is provided.
+    if config ~= nil then
+        treegen.write_file(self._dir, self._config_file_rel,
+                           yaml.encode(config))
+    end
+
+    -- Reload config on all the instances.
+    self:each(function(iserver)
+        -- Assume that all the instances are started.
+        --
+        -- This requirement may be relaxed if needed, it is just
+        -- for simplicity.
+        assert(iserver.process ~= nil)
+
+        iserver:exec(function()
+            local cfg = require('config')
+
+            cfg:reload()
+        end)
+    end)
+end
+
+local methods = {
+    each = cluster_each,
+    size = cluster_size,
+    start = cluster_start,
+    start_instance = cluster_start_instance,
+    stop = cluster_stop,
+    drop = cluster_drop,
+    sync = cluster_sync,
+    reload = cluster_reload,
+}
+
+local cluster_mt = {
+    __index = function(self, k)
+        if methods[k] ~= nil then
+            return methods[k]
+        end
+        if self._server_map[k] ~= nil then
+            return self._server_map[k]
+        end
+        return rawget(self, k)
+    end
+}
+
+--- Create a new Tarantool cluster.
+--
+-- @tab g Group of tests.
+-- @tab config Cluster configuration.
+-- @tab[opt] server_opts Extra options passed to server:new().
+-- @tab[opt] opts Cluster options.
+-- @string[opt] opts.dir Specific directory for the cluster.
+-- @return table
+local function new(g, config, server_opts, opts)
```

Comment:
1. Should we inherit from `object` for consistency with other luatest APIs?
2. Should we change it to `cluster:new()` (same reason)?
3. I would eliminate the test group argument like it was done for `treegen`.

Reply:
Hm, I used to avoid this approach but you're right it's better to remain consistent.

1. Applied.
2. Applied.
3. Applied.

Thank you.